### DISCUSSION TITLE
为什么 8 个字节的 UDP 能传输数据 · Why's THE Design? · whys-the-design-udp-minimum-header

### DISCUSSION THREAD

**Discussion by draveness** [draveness/blog-comments#586] [Announcements]
https://draveness.mewhys-the-design-udp-minimum-header

**Comment by leafduo** [comment]
长度和校验码可以省略，长度可以从 IP 的长度推算出来，校验码没有就没有呗，RFC 里也说不想算的话可以填0。

其他传输层协议有的有端口号的概念有的没有，我感觉主要是看是否需要在一个主机上跑多个使用这个协议的程序，像 ICMP 这种没必要跑多个就没有端口的概念，类似 TCP、UDP 这种需要跑多个的就需要端口号来区分。

**Comment by zhaolunallen** [comment]
看了rfc768的文档，很疑惑，明明长度可以通过ip推到出来，那为什么还要设计进去呢？

**Comment by draveness** [comment]
> 看了rfc768的文档，很疑惑，明明长度可以通过ip推到出来，那为什么还要设计进去呢？

你说的是数据报的长度通过 IP Packet 长度推导出来么

**Comment by zhaolunallen** [comment]
是的，upd的数据包长度可以通过ip Packet 长度推导出来（看资料说tcp包长度也是通过ip 包长度推导出来的），那为什么udp不跟tcp保持一致，而要搞一个冗余的长度字段呢？  看了下 tcp-ip详解里，卷1提到了是upd长度是冗余的，卷2里原话提到了“为什么存在 UDP 长度字段呢?可能是为了加上少量的差错检测，因为 U D P 检验和是可选的。”这句话，没明白它的意思，求大佬解惑。

**Comment by draveness** [comment]
> 是的，upd的数据包长度可以通过ip Packet 长度推导出来（看资料说tcp包长度也是通过ip 包长度推导出来的），那为什么udp不跟tcp保持一致，而要搞一个冗余的长度字段呢？ 看了下 tcp-ip详解里，卷1提到了是upd长度是冗余的，卷2里原话提到了“为什么存在 UDP 长度字段呢?可能是为了加上少量的差错检测，因为 U D P 检验和是可选的。”这句话，没明白它的意思，求大佬解惑。

找到两种说法，都可以解释这个问题

1. 除了 IP 协议还有其他的协议可以传输 UDP 包[1]
2. UDP 的内容是自包含的，TCP 一般都是都是数据流的一部分[2]

[1]: Why UDP header has 'length field'? <https://stackoverflow.com/questions/41394281/why-udp-header-has-length-field>
[2]: Why UDP has a TOTAL LENGTH field but TCP does not!? https://kplug-list.kernel-panic.narkive.com/OXz0n7TG/why-udp-has-a-total-length-field-but-tcp-does-not

**Comment by nopoetry** [comment]
大佬, 关于UDP长度和校验和字段是否可以被省略这个问题, 
在<<计算机网络自顶向下>>中还提到了, UDP之所以提供了'冗余'的端到端差错检验, 是因为并不是所有链路层都会进行差错校验.

长度是否可以被省略, 除了您提到的 `除了 IP 协议还有其他的协议可以传输 UDP 包[1]` , 还有就是把长度作为校验和的一部分, 用来进行差错校验, 所以不能被省略

校验和是否可以被省略, 并不是所有链路层都会进行差错校验, 所以应该也不能被省略

不知道我这样理解对不对, 还请您指正

**Comment by kedeup** [comment]
> @nopoetry
> 大佬, 关于UDP长度和校验和字段是否可以被省略这个问题, 
> 在<<计算机网络自顶向下>>中还提到了, UDP之所以提供了'冗余'的端到端差错检验, 是因为并不是所有链路层都会进行差错校验.
> 
> 长度是否可以被省略, 除了您提到的 `除了 IP 协议还有其他的协议可以传输 UDP 包[1]` , 还有就是把长度作为校验和的一部分, 用来进行差错校验, 所以不能被省略
> 
> 校验和是否可以被省略, 并不是所有链路层都会进行差错校验, 所以应该也不能被省略
> 
> 不知道我这样理解对不对, 还请您指正

即使把长度作为校验和的一部分，那也是可以推算出来的啊，这并不是长度可以省略的原因

**Comment by hehehuai** [comment]
udp不能省略长度和校验，一个udp报文可能会超过mtu的长度会分片。
